airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/global_substream_cursor.py

```diff
@@ -0,0 +1,283 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import threading
+import time
+from typing import Any, Iterable, Mapping, Optional, Union, Generator
+
+from airbyte_cdk.sources.declarative.incremental.datetime_based_cursor import DatetimeBasedCursor
+from airbyte_cdk.sources.declarative.incremental.declarative_cursor import DeclarativeCursor
+from airbyte_cdk.sources.declarative.partition_routers.partition_router import PartitionRouter
+from airbyte_cdk.sources.types import Record, StreamSlice, StreamState
+
+
+class Timer:
+    """
+    A simple timer class that measures elapsed time in seconds using a high-resolution performance counter.
+    """
+
+    def __init__(self) -> None:
+        self._start: Optional[int] = None
+
+    def start(self) -> None:
+        self._start = time.perf_counter_ns()
+
+    def finish(self) -> int:
+        if self._start:
+            return int((time.perf_counter_ns() - self._start) // 1e9)
+        else:
+            raise RuntimeError("Global substream cursor timer not started")
+
+
+class GlobalSubstreamCursor(DeclarativeCursor):
+    """
+    The GlobalSubstreamCursor is designed to track the state of substreams using a single global cursor.
+    This class is beneficial for streams with many partitions, as it allows the state to be managed globally
+    instead of per partition, simplifying state management and reducing the size of state messages.
+
+    This cursor is activated by setting the `global_substream_cursor` parameter for incremental sync.
+
+    Warnings:
+    - This class enforces a minimal lookback window for substream based on the duration of the previous sync to avoid losing records. This lookback ensures that any records added or updated during the sync are captured in subsequent syncs.
+    - The global cursor is updated only at the end of the sync. If the sync ends prematurely (e.g., due to an exception), the state will not be updated.
+    - When using the `incremental_dependency` option, the sync will progress through parent records, preventing the sync from getting infinitely stuck. However, it is crucial to understand the requirements for both the `global_substream_cursor` and `incremental_dependency` options to avoid data loss.
+    """
+
+    def __init__(self, stream_cursor: DatetimeBasedCursor, partition_router: PartitionRouter):
+        self._stream_cursor = stream_cursor
+        self._partition_router = partition_router
+        self._timer = Timer()
+        self._lock = threading.Lock()
```

Comment:
@brianjlai can you confirm what the plan is for the coalescing of the cursors? Will we be using the existing low-code classes with the concurrent cdk?

context: this is adding thread safety logic in case this does get use in a concurrent context

Reply:
So our current plan is that the low-code processing (the entrypoint being `DeclarativeStream.read_records()` won't be managing state or cursors at all. And the concurrent processing framework will be responsible for instantiating and invoking cursor methods. So right now that means using the existing `ConcurrentCursor`.

As it pertains to this work, this seems like it should be fine. But I think whenever we get to substream state for concurrent, we would need to port the new low-code global cursor into concurrent because there is no existing substream concurrent cursor implementation